File: app/src/main/java/desafio/com/br/github/ui/pull_request/PullRequestActivity.java
package desafio.com.br.github.ui.pull_request;

import android.app.*;
import android.os.*;
import android.support.v7.widget.*;
import java.util.*;
import butterknife.*;
import desafio.com.br.github.R;
import desafio.com.br.github.data.network.model.pull_request.*;


/**
 * Created by rafael on 24/01/18.
 */


public class PullRequestActivity extends Activity implements IviewPullRequest{
    @BindView(R.id.repositories_list)
    RecyclerView repositoriesRecyclerView;
    IpresenterPullRequest.PresenterView presenterPullRequest;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        getActionBar().setDisplayHomeAsUpEnabled(true);


        presenterPullRequest = new PresenterPullRequest(this);
        presenterPullRequest.fetchData(getIntent().getStringExtra("creator"),getIntent().getStringExtra("repository"));

        ButterKnife.bind(this);
    }

    @Override
    public void Success(ArrayList<PullRequest> pullRequests) {

        LinearLayoutManager layoutManager = new LinearLayoutManager(PullRequestActivity.this);
        repositoriesRecyclerView.setLayoutManager(layoutManager);
        pullRequestAdapter adapter = new pullRequestAdapter(pullRequests,PullRequestActivity.this);
        repositoriesRecyclerView.setAdapter(adapter);
    }

    @Override
    public void Failure(String error) {

    }
}

File: app/src/main/java/com/valderas/leonardo/nexaasdesafio/main/mvvm/view/pull/PullRequestActivity.kt
package com.valderas.leonardo.nexaasdesafio.main.mvvm.view.pull

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import android.webkit.URLUtil
import com.valderas.leonardo.nexaasdesafio.R
import com.valderas.leonardo.nexaasdesafio.main.mvvm.model.Contributor
import com.valderas.leonardo.nexaasdesafio.main.mvvm.model.Repo
import com.valderas.leonardo.nexaasdesafio.main.mvvm.view.main.MainAdapter
import com.valderas.leonardo.nexaasdesafio.main.mvvm.viewmodel.MainViewModel
import com.valderas.leonardo.nexaasdesafio.main.mvvm.viewmodel.PullRequestViewModel
import com.valderas.leonardo.nexaasdesafio.main.utils.EndlessRecyclerOnScrollListener
import com.valderas.leonardo.nexaasdesafio.main.utils.GenericEntityTransactionState
import com.valderas.leonardo.nexaasdesafio.main.utils.Helpers
import com.valderas.leonardo.nexaasdesafio.main.utils.Page
import dagger.android.support.DaggerAppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.toolbar.*
import javax.inject.Inject

class PullRequestActivity : DaggerAppCompatActivity(), PullRequestAdapter.OnItemClickListener {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    lateinit var viewModel: PullRequestViewModel

    var recyclerAdapter = PullRequestAdapter(mutableListOf(), this)

    private var page = Page()


    private var name = ""
    private var owner = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        init()
    }

    private fun init(){
        getStringsIntentExtra()
        initViewModel()
        initRecyclerView()
        getRepoListFromViewModel()
    }

    private fun initViewModel() {
        viewModel = ViewModelProviders.of(this, viewModelFactory)
                .get(PullRequestViewModel::class.java)

        viewModel.pullStateLive.observe(this,
                Observer<GenericEntityTransactionState> {
                    it?.let {
                        when (it) {
                            is GenericEntityTransactionState.IsLoading -> {
                                pb_loading.visibility = if (it.isloading) View.VISIBLE else View.GONE
                            }
                            is GenericEntityTransactionState.IsSuccess<*> -> {
                                @Suppress("UNCHECKED_CAST")
                                recyclerAdapter.rangeItems(it.entity as MutableList<Contributor>)
                            }
                            is GenericEntityTransactionState.Error -> {
                                it.idResource
                                // showMessage(it.msg)
                            }
                        }
                    }
                })
    }

    fun getStringsIntentExtra() {
        owner = intent.getStringExtra(Helpers.OWNER_EXTRA)
        name = intent.getStringExtra(Helpers.NAME_EXTRA)
    }

    override fun onItemClick(url: String, view: View) {
        if (URLUtil.isValidUrl(url))
            startImpliciteIntent(url)
        else
            error("Url invalida.")
    }

    private fun startImpliciteIntent(url: String) {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse(url)
        startActivity(intent)
    }

    private fun initRecyclerView() {
        with(recycler_view) {
            layoutManager = LinearLayoutManager(context)
            setHasFixedSize(false)
            adapter = recyclerAdapter
            clearOnScrollListeners()
            addOnScrollListener(object : EndlessRecyclerOnScrollListener() {
                override fun scrollUp(isUp: Boolean) {
                    //for buttom float
                }

                override fun onLoadMore() {
                    page.pageNumber++
                    getRepoListFromViewModel()
                }
            })
        }
    }

    private fun getRepoListFromViewModel() = viewModel.getPullRequestList(owner, name, page)
}
